lib/pstore.rb

```diff
@@ -10,87 +10,318 @@
 
 require "digest"
 
-#
-# PStore implements a file based persistence mechanism based on a Hash.  User
-# code can store hierarchies of Ruby objects (values) into the data store file
-# by name (keys).  An object hierarchy may be just a single object.  User code
-# may later read values back from the data store or even update data, as needed.
+# \PStore implements a file based persistence mechanism based on a Hash.
+# User code can store hierarchies of Ruby objects (values)
+# into the data store by name (keys).
+# An object hierarchy may be just a single object.
+# User code may later read values back from the data store
+# or even update data, as needed.
 #
 # The transactional behavior ensures that any changes succeed or fail together.
-# This can be used to ensure that the data store is not left in a transitory
-# state, where some values were updated but others were not.
+# This can be used to ensure that the data store is not left in a transitory state,
+# where some values were updated but others were not.
+#
+# Behind the scenes, Ruby objects are stored to the data store file with Marshal.
+# That carries the usual limitations. Proc objects cannot be marshalled,
+# for example.
+#
+# There are three key terms here (details at the links):
+#
+# - {Store}[rdoc-ref:PStore@The+Store]: a store is an instance of \PStore.
+# - {Roots}[rdoc-ref:PStore@Roots]: the store is hash-like;
+#   each root is a key for a stored object.
+# - {Transactions}[rdoc-ref:PStore@Transactions]: each transaction is a collection
+#   of prospective changes to the store;
+#   a transaction is defined in the block given with a call
+#   to PStore#transaction.
+#
+# == About the Examples
+#
+# Examples on this page need a store that has known properties.
+# Therefore examples may assume the existence of this helper method:
+#
+#   require 'pstore'
+#   require 'tempfile'
+#   # Yield a pristine store for use in examples.
+#   def example_store
+#     # Create the store in a temporary file.
+#     path = Tempfile.create
+#     store = PStore.new(path)
+#     # Populate the store.
+#     store.transaction do
+#       store[:foo] = 0
+#       store[:bar] = 1
+#       store[:baz] = 2
+#     end
+#     yield store
```

Comment:
I find `example_store` yielding a block is a little odd. Maybe it should just return the store instead?

Reply:
But this isn't automatically cleaning up the tempfile anyways. In order to do that, you have to do

```ruby
def example_store
  # Create the store in a temporary file.
  Tempfile.create do |file|
    store = PStore.new(file)

    # Populate the store.
    store.transaction do
      store[:foo] = 0
      store[:bar] = 1
      store[:baz] = 2
    end

    yield store
  end
end
```

Reply:
Lemme put the use of `example_store` first, then its implementation.

Reply:
Is that how you want it?

Reply:
I need a reliably-populated store for the examples.  The duped store (if that's what you mean by 'before') modifies the store.

To go static, I'd have to create and populate locally in each example.

Reply:
I see, that's a fair argument.

Reply:
So I'll adopt your corrected example_store`.

Reply:
Adopted.